### PR TITLE
Separate reduce functions from math

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary:
Separate CPU reduce functions from math

i-am-not-moving-c2-to-c10

Differential Revision: D13999469

### CODE REVIEW

File: caffe2/utils/math_test.cc [hunk 1]
```diff
@@ -426,253 +426,6 @@ TEST(MathTest, FloatToHalfConversion) {
 
 namespace {
 
-class ReduceTensorTest : public testing::Test {
```
Comment (yinghai) [inline caffe2/utils/math_test.cc]: Where do these tests go?

Comment (BIT-silence) [reply caffe2/utils/math_test.cc]: This part is actually covered by reduce_ops_test. So there is no need to keep such cc_tests.

File: caffe2/utils/math/reduce.h [hunk 2]
```diff
@@ -5,8 +5,85 @@
...
+ReduceMin(const int N, const T* X, T* y, Tensor* scratch_ptr, Context* context);
+
+template <typename T, class Context>
+CAFFE2_API void
+ReduceMax(const int N, const T* X, T* y, Tensor* scratch_ptr, Context* context);
+
+// Y = alpha * ReduceMin(X)
+template <typename T, class Context>
+CAFFE2_API void ReduceMin(
+    const int ndim,
```
Comment (houseroad) [inline caffe2/utils/math/reduce.h]: Explicitly mention that X_dims and Y_dims share the same ndim?

File: caffe2/utils/math/reduce.h [hunk 3]
```diff
@@ -5,8 +5,85 @@
...
 
+template <typename T, class Context>
+CAFFE2_API void
+ReduceMin(const int N, const T* X, T* y, Tensor* scratch_ptr, Context* context);
+
+template <typename T, class Context>
+CAFFE2_API void
+ReduceMax(const int N, const T* X, T* y, Tensor* scratch_ptr, Context* context);
+
+// Y = alpha * ReduceMin(X)
```
Comment (houseroad) [inline caffe2/utils/math/reduce.h]: thumb for the comment :-)

File: caffe2/utils/math/reduce.h [hunk 4]
```diff
@@ -19,6 +96,7 @@ CAFFE2_API void Moments(
     Context* context);
 
 } // namespace math
+
```
Comment (houseroad) [inline caffe2/utils/math/reduce.h]: is this expected?

**Review by houseroad — APPROVED** [review]
Looks good.

File: caffe2/utils/math/reduce.h (hunk 4 above)
Comment (BIT-silence) [reply caffe2/utils/math/reduce.h]: Yes, this } is the end of namespace math.

File: caffe2/utils/math/reduce.h (hunk 2 above)
Comment (BIT-silence) [reply caffe2/utils/math/reduce.h]: Done

**Pull request closed without merging**